porch/controllers/functiondiscovery/controller.go

```diff
@@ -0,0 +1,183 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Options struct {
+}
+
+func (o *Options) InitDefaults() {
+}
+
+func (o *Options) BindFlags(prefix string, flags *flag.FlagSet) {
+}
+
+// FunctionReconciler creates Function objects
+type FunctionReconciler struct {
+	Options
+
+	client     client.Client
+	restConfig *rest.Config
+
+	ociStorage *kptoci.Storage
+}
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/controller-gen@v0.8.0 rbac:roleName=porch-controllers-functions webhook paths="." output:rbac:artifacts:config=../../../config/rbac
+
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=repositories,verbs=get;list;watch
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=functions,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=functions/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=functions/finalizers,verbs=update
+
+// Reconcile implements the main kubernetes reconciliation loop.
+func (r *FunctionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var subject api.Repository
+	if err := r.client.Get(ctx, req.NamespacedName, &subject); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var result ctrl.Result
+
+	applyResults, err := r.reconcileRepository(ctx, &subject)
+	if err != nil {
+		// TODO: Update status?
+		// TODO: Post event?
+		return result, err
+	}
+
+	if applyResults != nil && !(applyResults.AllApplied() && applyResults.AllHealthy()) {
+		result.Requeue = true
+		return result, nil
+	}
+
+	return result, nil
+}
+
+func (r *FunctionReconciler) reconcileRepository(ctx context.Context, subject *api.Repository) (*applyset.ApplyResults, error) {
+	// TODO: Cache dynamicClient / discovery etc
+	restConfig := r.restConfig
+
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new dynamic client: %w", err)
+	}
+
+	// TODO: Use a better discovery client
+	discovery, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error building discovery client: %w", err)
+	}
+
+	cached := memory.NewMemCacheClient(discovery)
+
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(cached)
+
+	objects, err := r.buildObjectsToApply(ctx, subject)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Cache applyset
+	patchOptions := metav1.PatchOptions{
+		FieldManager: "functions-" + subject.GetNamespace() + "-" + subject.GetName(),
+	}
+
+	// We force to overcome errors like: Apply failed with 1 conflict: conflict with "kubectl-client-side-apply" using apps/v1: .spec.template.spec.containers[name="porch-server"].image
+	// TODO: How to handle this better
+	force := true
+	patchOptions.Force = &force
+
+	applyset, err := applyset.New(applyset.Options{
```

Comment:
note to self: need to learn about applyset :)

Reply:
Yeah, I like it ... but I would :-)